### PR TITLE
doc: improve documentation for create-ocp-project

### PR DESCRIPTION
**Pull request by evanshortiss**

### CODE REVIEW

File: docs/main/create-ocp-project/README.md [hunk 1]
```diff
@@ -23,88 +24,170 @@ TARGET_NS=sonataflow-infra
 ```
 
 # Installation
-## Persistence pre-requisites
-If persistence is enabled, you must have a PostgreSQL instance running in the cluster, in the same `namespace` as the workflows.
 
-A `secret` containing the instance credentials must exists as well. 
+## Persistence Pre-requisites
```
Comment (evanshortiss) [inline docs/main/create-ocp-project/README.md:28]: Is this relevant to the workflow, or can it be removed? Does the workflow talk directly to the database?

File: docs/main/create-ocp-project/README.md [hunk 2]
```diff
@@ -23,88 +24,170 @@ TARGET_NS=sonataflow-infra
...
+| `JIRA_URL`      | The Jira server URL, e.g `https://your-instance.atlassian.net` | ✅ | |
+| `JIRA_USERNAME`      | The Jira username, e.g `you@company.com` | ✅ | |
+| `JIRA_API_TOKEN`      | The Jira API token associated with the username. If you're using Jira cloud, you can obtain a token [at this link](https://id.atlassian.com/manage-profile/security/api-tokens). | ✅ | |
+| `OCP_API_SERVER_URL`      | The OpenShift API server url, e.g `https://api.cluster.hostname.com:6443` | ✅ | |
+| `OCP_API_SERVER_TOKEN`      | The authorization bearer token to use when sending request to OpenShift | ✅ | |
+
+To obtain an OpenShift API token, create a Service Account, assign permissions to it, and request a token:
+
+```bash
+oc create sa orchestrator-ocp-api
```
Comment (evanshortiss) [inline docs/main/create-ocp-project/README.md:63]: @gabriel-farache can you confirm the SA permissions? I tried `admin` and `self-provisioner` roles, but those failed. I had to use `cluster-admin` to get the project creation to work.

**Comment by pkliczewski** [comment]
@evanshortiss thank you for your contribution

File: docs/main/create-ocp-project/README.md (hunk 2 above)
Comment (gabriel-farache) [reply docs/main/create-ocp-project/README.md:55]: It's interesting that `admin` cluster role failed, when checking why:
```
$ oc adm policy add-cluster-role-to-user admin  -z orchestrator-ocp-api3
clusterrole.rbac.authorization.k8s.io/admin added: "orchestrator-ocp-api3"

$ oc auth can-i get project --as=system:serviceaccount:rhdh-operator:orchestrator-ocp-api3
Warning: resource 'projects' is not namespace scoped in group 'project.openshift.io'

yes

$oc auth can-i create project --as=system:serviceaccount:rhdh-operator:orchestrator-ocp-api3
Warning: resource 'projects' is not namespace scoped in group 'project.openshift.io'

no
```

In the workflow we are creating then reading the project created to get its state.
To create the project we are sending a POST request to ` /apis/project.openshift.io/v1/projects`, and we can see that the `admin` cannot `create` a `project`

But
```
$ oc auth can-i create projectrequest --as=system:serviceaccount:rhdh-operator:orchestrator-ocp-api3
Warning: resource 'projectrequests' is not namespace scoped in group 'project.openshift.io'

yes
```
It appears that what we are doing is wrong, instead, we should send a POST request to ` /apis/project.openshift.io/v1/projectRequests` for which the `admin` and `self-provisioner` cluster roles are set to allow its creation.
I will update the workflow to stick to the OCP way of creating project.

FYI `self-provisioner` only is not enough:
```
$ oc adm policy add-cluster-role-to-user self-provisioner  -z orchestrator-ocp-api4
clusterrole.rbac.authorization.k8s.io/self-provisioner added: "orchestrator-ocp-api4"

$ oc auth can-i create projectrequest --as=system:serviceaccount:rhdh-operator:orchestrator-ocp-api4
Warning: resource 'projectrequests' is not namespace scoped in group 'project.openshift.io'

yes

$ oc auth can-i get project --as=system:serviceaccount:rhdh-operator:orchestrator-ocp-api4
Warning: resource 'projects' is not namespace scoped in group 'project.openshift.io'

no
```

We need the admin:
```
$ oc adm policy add-cluster-role-to-user admin  -z orchestrator-ocp-api4
clusterrole.rbac.authorization.k8s.io/admin added: "orchestrator-ocp-api4"

$ oc auth can-i get project --as=system:serviceaccount:rhdh-operator:orchestrator-ocp-api4
Warning: resource 'projects' is not namespace scoped in group 'project.openshift.io'

yes
```

Using the `cluster-admin` will work but it will be able to do anything on the cluster, whereas the `admin` may have more restricted set of permissions

Alternatively, we could add a step to create a cluster role to get the project
```
$oc create clusterrole read-project --verb=get --resource=projects
clusterrole.rbac.authorization.k8s.io/read-project created

$ oc adm policy add-cluster-role-to-user read-project  -z orchestrator-ocp-api4
clusterrole.rbac.authorization.k8s.io/read-project added: "orchestrator-ocp-api4"

$ oc auth can-i get project --as=system:serviceaccount:rhdh-operator:orchestrator-ocp-api4
Warning: resource 'projects' is not namespace scoped in group 'project.openshift.io'

yes
```

**Review by gabriel-farache — CHANGES_REQUESTED** [review]
Very nice, thx
I reply to your question about the permissions 
1 little change: we should store or use the output of the create token command

File: docs/main/create-ocp-project/README.md (hunk 1 above)
Comment (gabriel-farache) [reply docs/main/create-ocp-project/README.md]: This shall be kept, the workflow is persisting its state to the DB directly

File: docs/main/create-ocp-project/README.md [hunk 3]
```diff
@@ -14,97 +14,176 @@ The following table lists the configurable parameters of the Create OCP project
...
+> If you run the `helm upgrade` command, the values of the Secret are reset.
+
+Run the following command to do so. Replace the example values with the
+correct values for your environment:
+
+```bash
+export TARGET_NS='sonataflow-infra'
+export WORKFLOW_NAME='create-ocp-project'
+
+export NOTIFICATIONS_BEARER_TOKEN='token_from_backstage_appconfig'
```
Comment (gabriel-farache) [inline docs/main/create-ocp-project/README.md:79]: While we are at it, let's get the token directly:
```
export NOTIFICATIONS_BEARER_TOKEN=$(oc get secrets -n rhdh-operator backstage-backend-auth-secret -o go-template='{{ .data.BACKEND_SECRET  }}')
```

File: docs/main/create-ocp-project/README.md [hunk 4]
```diff
@@ -14,97 +14,176 @@ The following table lists the configurable parameters of the Create OCP project
...
+export BACKSTAGE_NOTIFICATIONS_URL='https://backstage.cluster.com'
```
Comment (gabriel-farache) [inline docs/main/create-ocp-project/README.md:80]: I would keep using `http://backstage-backstage.rhdh-operator` as value as this will be the value as long as the user follows the orchestrator README, in case it's not the case, it will have to update this anyway

File: docs/main/create-ocp-project/README.md [hunk 5]
```diff
@@ -14,97 +14,176 @@ The following table lists the configurable parameters of the Create OCP project
...
+      \"NOTIFICATIONS_BEARER_TOKEN\": \"$NOTIFICATIONS_BEARER_TOKEN\" \
+      \"JIRA_API_TOKEN\": \"$JIRA_API_TOKEN\", \
+      \"OCP_API_SERVER_TOKEN\": \"$OCP_API_SERVER_TOKEN\", \
+      \"BACKSTAGE_NOTIFICATIONS_URL\": \"$BACKSTAGE_NOTIFICATIONS_URL\", \
+      \"JIRA_URL\": \"$JIRA_URL\", \
+      \"JIRA_USERNAME\": \"$JIRA_USERNAME\", \
+      \"OCP_API_SERVER_URL\": \"$OCP_API_SERVER_URL\", \
+      \"OCP_CONSOLE_URL\": \"$OCP_CONSOLE_URL\" \
+    } \
+  }"
```
Comment (gabriel-farache) [inline docs/main/create-ocp-project/README.md:102]: This should be
```
oc -n $TARGET_NS patch secret "$WORKFLOW_NAME-creds" \
  --type merge -p "{
    \"data\": {
        \"NOTIFICATIONS_BEARER_TOKEN\": \"$NOTIFICATIONS_BEARER_TOKEN\"
    },
    \"stringData\": { 
      \"JIRA_API_TOKEN\": \"$JIRA_API_TOKEN\", 
      \"OCP_API_SERVER_TOKEN\": \"$OCP_API_SERVER_TOKEN\", 
      \"BACKSTAGE_NOTIFICATIONS_URL\": \"$BACKSTAGE_NOTIFICATIONS_URL\", 
      \"JIRA_URL\": \"$JIRA_URL\", 
      \"JIRA_USERNAME\": \"$JIRA_USERNAME\", 
      \"OCP_API_SERVER_URL\": \"$OCP_API_SERVER_URL\", 
      \"OCP_CONSOLE_URL\": \"$OCP_CONSOLE_URL\" 
    }
  }"
```

File: docs/main/create-ocp-project/README.md (hunk 2 above)
Comment (gabriel-farache) [reply docs/main/create-ocp-project/README.md:55]: cc @ElaiShalevRH

File: docs/main/create-ocp-project/README.md (hunk 4 above)
Comment (evanshortiss) [reply docs/main/create-ocp-project/README.md:80]: IMO we shouldn't assume the user is deploying Backstage using Orchestrator. Most people will be adding Orchestrator to an existing Backstage instance that is unlikely to match `http://backstage-backstage.rhdh-operator`.

Comment (evanshortiss) [reply docs/main/create-ocp-project/README.md:80]: Maybe I can change `cluster` to `replace-me` to make it more obvious

File: docs/main/create-ocp-project/README.md (hunk 5 above)
Comment (evanshortiss) [reply docs/main/create-ocp-project/README.md:102]: This assumes the Backstage instance is deployed with the Orchestrator Operator. I think the instructions should be more generic, so if someone has deployed using the RHDH Operator/Helm chart we can tell them where to look. WDYT/

Comment (evanshortiss) [reply docs/main/create-ocp-project/README.md:102]: For example, there's no `backstage-backend-auth-secret` if you're using the official operator. Instead this value is set like so:

Screenshot 2024-11-04 at 10 46 38 AM

File: docs/main/create-ocp-project/README.md (hunk 3 above)
Comment (evanshortiss) [reply docs/main/create-ocp-project/README.md:79]: @gabriel-farache I added comments in another thread here to discuss this.

File: docs/main/create-ocp-project/README.md (hunk 2 above)
Comment (evanshortiss) [reply docs/main/create-ocp-project/README.md:55]: > Alternatively, we could add a step to create a cluster role to get the project

You mean a cluster role that can create/list projects?

Comment (gabriel-farache) [reply docs/main/create-ocp-project/README.md:55]: > >Alternatively, we could add a step to create a cluster role to get the project

> You mean a cluster role that can create/list projects?

No, just `get` a project: in the workflow, once the project is created, we retrieve the project to get its status.

We do not need to `list` all of the projects as we are not checking if a project with similar name exists.
And we do not need `create` either as https://github.com/parodos-dev/serverless-workflows/pull/490 introduced a change to use `ProjectRequest` kind when creating a project instead of using a `POST` on the `Project` API.

If we want to use this alternative ( creating a cluster role), the SA will have to get assigned to `self-provisioner` (which provides the permissions to `create` a `ProjectRequest`)  and the newly create cluster role (which would provides the `get` permission on `Project`)

File: docs/main/create-ocp-project/README.md (hunk 4 above)
Comment (gabriel-farache) [reply docs/main/create-ocp-project/README.md:82]: > Most people will be adding Orchestrator to an existing Backstage instance

yeah, ok, I see your point

> Maybe I should change cluster to replace-me to make it more obvious?

Yes, that would be clearer and will catch the eye more easily

File: docs/main/create-ocp-project/README.md (hunk 5 above)
Comment (gabriel-farache) [reply docs/main/create-ocp-project/README.md:104]: > This assumes the Backstage instance is deployed with the Orchestrator Operator. I think the instructions should be more generic, so if someone has deployed using the RHDH Operator/Helm chart

That's a valid point: currently we take as assumption that the RHDH was installed using the Orchestrator Operator, @pkliczewski should we keep doing that?

> we can tell them where to look.

Yes, we can have 2 section: 

1. RHDH installed with RHDH operator: in this section we have 
```
export NOTIFICATIONS_BEARER_TOKEN=$(oc get secrets -n rhdh-operator backstage-backend-auth-secret -o go-template='{{ .data.BACKEND_SECRET  }}' | base64 -d)
```
Not the ` | base64 -d` , this way all values will be going to `stringData` in the secret's patch
2. RHDH not installed with RHDH operator: in this section we have instructions on where to look to get the token and then
```
export NOTIFICATIONS_BEARER_TOKEN='replace_me'
```

The rest should not change:
```
oc -n $TARGET_NS patch secret "$WORKFLOW_NAME-creds" \
  --type merge -p "{
    \"stringData\": { 
       \"NOTIFICATIONS_BEARER_TOKEN\": \"$NOTIFICATIONS_BEARER_TOKEN\",
      \"JIRA_API_TOKEN\": \"$JIRA_API_TOKEN\", 
      \"OCP_API_SERVER_TOKEN\": \"$OCP_API_SERVER_TOKEN\", 
      \"BACKSTAGE_NOTIFICATIONS_URL\": \"$BACKSTAGE_NOTIFICATIONS_URL\", 
      \"JIRA_URL\": \"$JIRA_URL\", 
      \"JIRA_USERNAME\": \"$JIRA_USERNAME\", 
      \"OCP_API_SERVER_URL\": \"$OCP_API_SERVER_URL\", 
      \"OCP_CONSOLE_URL\": \"$OCP_CONSOLE_URL\" 
    }
  }"
```

WDYT?

Comment (evanshortiss) [reply docs/main/create-ocp-project/README.md:104]: Sounds good 👍

File: docs/main/create-ocp-project/README.md [hunk 6]
```diff
@@ -14,97 +14,189 @@ The following table lists the configurable parameters of the Create OCP project
...
-If you are using Jira cloud, you can generate the `JIRA_API_TOKEN` using https://id.atlassian.com/manage-profile/security/api-tokens 
 
-The `OCP_API_SERVER_TOKEN` should be associated with a service account.
+Now, patch the Secret with these values:
+
+```bash
+oc -n $TARGET_NS patch secret "$WORKFLOW_NAME-creds" \
+  --type merge -p "{ \
+    \"stringData\": { \
+      \"NOTIFICATIONS_BEARER_TOKEN\": \"$NOTIFICATIONS_BEARER_TOKEN\" \
```
Comment (gabriel-farache) [inline docs/main/create-ocp-project/README.md:106]: You are missing a comma in the end here

File: docs/main/create-ocp-project/README.md [hunk 7]
```diff
@@ -14,97 +14,189 @@ The following table lists the configurable parameters of the Create OCP project
...
+      \"JIRA_API_TOKEN\": \"$JIRA_API_TOKEN\", \
+      \"OCP_API_SERVER_TOKEN\": \"$OCP_API_SERVER_TOKEN\", \
+      \"BACKSTAGE_NOTIFICATIONS_URL\": \"$BACKSTAGE_NOTIFICATIONS_URL\", \
+      \"JIRA_URL\": \"$JIRA_URL\", \
+      \"JIRA_USERNAME\": \"$JIRA_USERNAME\", \
+      \"OCP_API_SERVER_URL\": \"$OCP_API_SERVER_URL\", \
+      \"OCP_CONSOLE_URL\": \"$OCP_CONSOLE_URL\" \
+    } \
```
Comment (gabriel-farache) [inline docs/main/create-ocp-project/README.md:114]: There is no need for the `\` inside the json string in the end of the line:
```
... -p "{ 
    \"stringData\": { 
      \"NOTIFICATIONS_BEARER_TOKEN\": \"$NOTIFICATIONS_BEARER_TOKEN\",
      \"JIRA_API_TOKEN\": \"$JIRA_API_TOKEN\", 
      \"OCP_API_SERVER_TOKEN\": \"$OCP_API_SERVER_TOKEN\", 
      \"BACKSTAGE_NOTIFICATIONS_URL\": \"$BACKSTAGE_NOTIFICATIONS_URL\", 
      \"JIRA_URL\": \"$JIRA_URL\", 
      \"JIRA_USERNAME\": \"$JIRA_USERNAME\", 
      \"OCP_API_SERVER_URL\": \"$OCP_API_SERVER_URL\", 
      \"OCP_CONSOLE_URL\": \"$OCP_CONSOLE_URL\" 
    }
  }"
```
works too

**Review by gabriel-farache — CHANGES_REQUESTED** [review]
one little comma to add and we are good to go

And please rebase :)

File: docs/main/create-ocp-project/README.md (hunk 6 above)
Comment (evanshortiss) [reply docs/main/create-ocp-project/README.md]: Nice find!

**Review by gabriel-farache — APPROVED** [review]
Thank you very much for your contribution :)
